service/kvmclusterconfig/v1/key/key_test.go

```diff
@@ -134,44 +147,46 @@ func Test_ToCustomObject(t *testing.T) {
 					},
 				},
 			},
-			ExpectedError: nil,
+			expectedError: nil,
 		},
-		// Failure : Nil *v1alpha1.KVMClusterConfig is not accepted.
 		{
-			InputObject:          emptyKVMClusterConfigPtr,
-			ExpectedCustomObject: v1alpha1.KVMClusterConfig{},
-			ExpectedError:        emptyValueError,
+			description:          "nil pointer to KVMClusterConfig must return emptyValueError",
+			inputObject:          emptyKVMClusterConfigPtr,
+			expectedCustomObject: v1alpha1.KVMClusterConfig{},
+			expectedError:        emptyValueError,
 		},
-		// Failure: Non-reference value of KVMClusterConfig is not accepted.
 		{
-			InputObject:          v1alpha1.KVMClusterConfig{},
-			ExpectedCustomObject: v1alpha1.KVMClusterConfig{},
-			ExpectedError:        wrongTypeError,
+			description:          "non-pointer value of KVMClusterConfig must return wrontTypeError",
+			inputObject:          v1alpha1.KVMClusterConfig{},
+			expectedCustomObject: v1alpha1.KVMClusterConfig{},
+			expectedError:        wrongTypeError,
 		},
-		// Failure: Wrong type must not be accepted
 		{
-			InputObject:          &v1alpha1.AzureClusterConfig{},
-			ExpectedCustomObject: v1alpha1.KVMClusterConfig{},
-			ExpectedError:        wrongTypeError,
+			description:          "wrong type must return wrongTypeError",
+			inputObject:          &v1alpha1.AzureClusterConfig{},
+			expectedCustomObject: v1alpha1.KVMClusterConfig{},
+			expectedError:        wrongTypeError,
 		},
-		// Failure: Nil interface{} must fail
 		{
-			InputObject:          nil,
-			ExpectedCustomObject: v1alpha1.KVMClusterConfig{},
-			ExpectedError:        wrongTypeError,
+			description:          "nil interface{} must return wrongTypeError",
+			inputObject:          nil,
+			expectedCustomObject: v1alpha1.KVMClusterConfig{},
+			expectedError:        wrongTypeError,
 		},
 	}
 
-	for i, tc := range testCases {
-		customObject, err := ToCustomObject(tc.InputObject)
-		if microerror.Cause(err) != tc.ExpectedError {
-			t.Errorf("TestCase %d: Received error %#v doesn't match expected %#v",
-				(i + 1), err, tc.ExpectedError)
-		}
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			customObject, err := ToCustomObject(tc.inputObject)
+			if microerror.Cause(err) != tc.expectedError {
+				t.Errorf("Received error %#v doesn't match expected %#v",
+					err, tc.expectedError)
+			}
 
-		if !reflect.DeepEqual(customObject, tc.ExpectedCustomObject) {
-			t.Errorf("TestCase %d: CustomObject %#v doesn't match expected %#v",
-				(i + 1), customObject, tc.ExpectedCustomObject)
-		}
+			if !reflect.DeepEqual(customObject, tc.expectedCustomObject) {
```

Comment:
I think there should be one thing tested in one sub test.

Reply:
I would appreciate an entry about that and running subtests in general in github.com/giantswarm/fmt. I'd like to see pros and cons of the approach in general and agree on it. I haven't notice this style being popular in the community (which in my case would be enough justification to learn and get used to that). My thinking is I more often care about reading test code than reading the test output, which I can adjust any time. So I would rather optimize test code outline than the test output outline.

\cc @fgimenez

Reply:
IMO there shouldn't be a hard bind between subtests and things being tested, if you are ok with testing multiple things in a test, you should be good with doing the same in a subtest.

As I understand https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks they help you organizing your tests, making it easy to run table-driven tests/benchmarks but also helping to give a hierarchy to your tests, with a direct way of expressing dependencies or setup and teardown steps common to several subtests. They also provide a built-in way to easily run groups of tests in parallel.

But I'm just rephrasing what is already said in the language docs :) Other than that from my side I find very useful to be able to filter by subtest with `-run`, and also to have a clear way of determining which iteration of your test cases' table failed, clear and included in the standard library.  

I'm not sure if we need to impose a way of doing things, if you feel ok with the way you are currently using it I think that should be fine. I just mentioned subtests and description fields in the previous PR because of the questions Tuomas was asking. Having different examples of writing tests, or code in general, in the code base is good IMO, it's not that hard to understand what's going on and it can give you ideas from other approaches, my 2c :)